File: ios-native/Sources/Supabase.swift
import Foundation

struct SupaError: LocalizedError {
    let message: String
    var errorDescription: String? { message }
}

/// Minimal Supabase client (Auth + PostgREST) built on URLSession — no SPM deps,
/// so the app builds on a bare CI runner. The anon/publishable key is public by
/// design (same as the website).
final class Supa {
    static let shared = Supa()

    // Public project config (matches the website's baked-in fallbacks).
    let baseURL = URL(string: "https://xbwhvkzgbnyqsjplbyox.supabase.co")!
    let anonKey = "sb_publishable_Jcs8IXK6YnOwWe-qYAFecg_Tb6M-ohu"

    private(set) var session: Session?
    private let sessionKey = "askai.session"

    private init() { loadSession() }

    // MARK: - Session persistence

    private func loadSession() {
        if let data = UserDefaults.standard.data(forKey: sessionKey),
           let s = try? JSONDecoder().decode(Session.self, from: data) {
            session = s
        }
    }
    private func saveSession(_ s: Session?) {
        session = s
        if let s, let data = try? JSONEncoder().encode(s) {
            UserDefaults.standard.set(data, forKey: sessionKey)
        } else {
            UserDefaults.standard.removeObject(forKey: sessionKey)
        }
    }

    var isAuthed: Bool { session != nil }
    var token: String { session?.accessToken ?? anonKey }
    var userId: String? { session?.userId }

    // MARK: - Auth

    func signIn(email: String, password: String) async throws -> Session {
        let s = try await authToken(grant: "password", body: ["email": email, "password": password])
        saveSession(s); return s
    }

    func signUp(email: String, password: String) async throws -> Session {
        // Auto-confirm is enabled, so signup usually returns a session directly.
        let url = baseURL.appendingPathComponent("auth/v1/signup")
        let data = try await send(url: url, method: "POST", auth: false,
                                  body: ["email": email, "password": password])
        if let s = try? parseSession(data, fallbackEmail: email) {
            saveSession(s); return s
        }
        // Confirmation required → try to sign in (may fail until confirmed).
        return try await signIn(email: email, password: password)
    }

    func refreshIfPossible() async {
        guard let rt = session?.refreshToken else { return }
        if let s = try? await authToken(grant: "refresh_token", body: ["refresh_token": rt]) {
            saveSession(s)
        }
    }

    func signOut() { saveSession(nil) }

    private func authToken(grant: String, body: [String: Any]) async throws -> Session {
        var comps = URLComponents(url: baseURL.appendingPathComponent("auth/v1/token"), resolvingAgainstBaseURL: false)!
        comps.queryItems = [URLQueryItem(name: "grant_type", value: grant)]
        let data = try await send(url: comps.url!, method: "POST", auth: false, body: body)
        return try parseSession(data, fallbackEmail: body["email"] as? String)
    }

    private func parseSession(_ data: Data, fallbackEmail: String?) throws -> Session {
        guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw SupaError(message: "Unexpected response")
        }
        if let msg = obj["error_description"] as? String ?? obj["msg"] as? String, obj["access_token"] == nil {
            throw SupaError(message: msg)
        }
        guard let at = obj["access_token"] as? String,
              let rt = obj["refresh_token"] as? String else {
            throw SupaError(message: "Could not start a session")
        }
        let user = obj["user"] as? [String: Any]
        let uid = (user?["id"] as? String) ?? ""
        let mail = (user?["email"] as? String) ?? fallbackEmail
        return Session(accessToken: at, refreshToken: rt, userId: uid, email: mail)
    }

    // MARK: - PostgREST

    /// GET rows. `path` is like "agents?workspace_id=eq.123&select=*&order=created_at".
    func select<T: Decodable>(_ path: String, as type: T.Type = T.self) async throws -> [T] {
        guard let full = URL(string: baseURL.absoluteString + "/rest/v1/" + path) else {
            throw SupaError(message: "Bad path")
        }
        let data = try await send(url: full, method: "GET", auth: true, body: nil)
        return try decodeArray(data)
    }

    @discardableResult
    func insert<T: Decodable>(_ table: String, _ row: [String: Any], returning: Bool = true, as type: T.Type = T.self) async throws -> [T] {
        let full = URL(string: baseURL.absoluteString + "/rest/v1/" + table)!
        let data = try await send(url: full, method: "POST", auth: true, body: row,
                                  prefer: returning ? "return=representation" : "return=minimal")
        if !returning { return [] }
        return try decodeArray(data)
    }

    func update(_ pathWithFilter: String, _ patch: [String: Any]) async throws {
        let full = URL(string: baseURL.absoluteString + "/rest/v1/" + pathWithFilter)!
        _ = try await send(url: full, method: "PATCH", auth: true, body: patch, prefer: "return=minimal")
    }

    func delete(_ pathWithFilter: String) async throws {
        let full = URL(string: baseURL.absoluteString + "/rest/v1/" + pathWithFilter)!
        _ = try await send(url: full, method: "DELETE", auth: true, body: nil, prefer: "return=minimal")
    }

    private func decodeArray<T: Decodable>(_ data: Data) throws -> [T] {
        if data.isEmpty { return [] }
        do { return try JSONDecoder().decode([T].self, from: data) }
        catch {
            // Surface PostgREST error bodies clearly.
            if let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let msg = obj["message"] as? String {
                throw SupaError(message: msg)
            }
            throw error
        }
    }

    // MARK: - Core request

    private func send(url: URL, method: String, auth: Bool, body: [String: Any]?, prefer: String? = nil) async throws -> Data {
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue(anonKey, forHTTPHeaderField: "apikey")
        req.setValue("Bearer \(auth ? token : anonKey)", forHTTPHeaderField: "Authorization")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let prefer { req.setValue(prefer, forHTTPHeaderField: "Prefer") }
        if let body { req.httpBody = try JSONSerialization.data(withJSONObject: body) }
        req.timeoutInterval = 30

        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw SupaError(message: "No response") }
        if http.statusCode == 401 && auth {
            // Token may have expired — refresh once and retry.
            await refreshIfPossible()
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            let (d2, r2) = try await URLSession.shared.data(for: req)
            if let h2 = r2 as? HTTPURLResponse, (200..<300).contains(h2.statusCode) { return d2 }
        }
        guard (200..<300).contains(http.statusCode) else {
            let msg = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["message"] as? String
                ?? String(data: data, encoding: .utf8) ?? "Request failed (\(http.statusCode))"
            throw SupaError(message: msg)
        }
        return data
    }
}

File: android/app/src/main/java/gg/askai/android/data/Supabase.kt
package gg.askai.android.data

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Minimal Supabase client (Auth + PostgREST + Storage) built on OkHttp — same
 * public project as the iOS app and website. The anon key is public by design.
 *
 * Reliability: the access token is refreshed proactively before it expires and
 * every REST call retries once after a forced refresh on 401, so writes no
 * longer fail silently when the session goes stale mid-use.
 */
object Supa {
    const val baseURL = "https://xbwhvkzgbnyqsjplbyox.supabase.co"
    const val anonKey = "sb_publishable_Jcs8IXK6YnOwWe-qYAFecg_Tb6M-ohu"

    private val http = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(120, TimeUnit.SECONDS)
        .build()
    private val JSON = "application/json".toMediaType()

    var accessToken: String? = null
    var refreshToken: String? = null
    var userId: String? = null
    var email: String? = null

    /** Epoch seconds when the access token expires (0 = unknown). */
    private var expiresAt: Long = 0L

    /** Set when the server rejects our refresh token — the user must sign in again. */
    @Volatile var sessionExpired = false
        private set

    private val refreshLock = Mutex()
    private lateinit var prefs: android.content.SharedPreferences

    fun init(ctx: Context) {
        prefs = ctx.getSharedPreferences("askai", Context.MODE_PRIVATE)
        accessToken = prefs.getString("access", null)
        refreshToken = prefs.getString("refresh", null)
        userId = prefs.getString("uid", null)
        email = prefs.getString("email", null)
        expiresAt = prefs.getLong("expires", 0L)
    }

    val isAuthed get() = accessToken != null && userId != null

    // Small app preferences that ride on the same store (theme etc.).
    var themeMode: String
        get() = prefs.getString("theme", "system") ?: "system"
        set(v) { prefs.edit().putString("theme", v).apply() }

    private fun persist() {
        prefs.edit()
            .putString("access", accessToken).putString("refresh", refreshToken)
            .putString("uid", userId).putString("email", email)
            .putLong("expires", expiresAt).apply()
    }

    fun signOut() {
        accessToken = null; refreshToken = null; userId = null; email = null
        expiresAt = 0L; sessionExpired = false; persist()
    }

    private fun authHeader() = "Bearer " + (accessToken ?: anonKey)

    // MARK: Auth

    /** Returns null on success, otherwise a human-readable error message. */
    suspend fun signIn(email: String, password: String) = auth("token?grant_type=password", email, password)
    suspend fun signUp(email: String, password: String) = auth("signup", email, password)

    private suspend fun auth(path: String, email: String, password: String): String? = withContext(Dispatchers.IO) {
        val body = JSONObject().put("email", email).put("password", password).toString()
        val req = Request.Builder()
            .url("$baseURL/auth/v1/$path")
            .addHeader("apikey", anonKey)
            .addHeader("Content-Type", "application/json")
            .post(body.toRequestBody(JSON)).build()
        try {
            http.newCall(req).execute().use { resp ->
                val txt = resp.body?.string() ?: return@withContext "Empty response from the server."
                val j = runCatching { JSONObject(txt) }.getOrDefault(JSONObject())
                if (!resp.isSuccessful) return@withContext authErrorMessage(j, resp.code)
                val at = j.optString("access_token", "")
                if (at.isEmpty()) return@withContext "Sign-in didn't return a session — try again."
                applySession(j)
                val user = j.optJSONObject("user")
                userId = user?.optString("id")
                this@Supa.email = user?.optString("email") ?: email
                sessionExpired = false
                persist(); null
            }
        } catch (e: IOException) {
            "Couldn't reach the server — check your internet connection."
        }
    }

    private fun authErrorMessage(j: JSONObject, code: Int): String {
        val msg = j.optString("error_description",
            j.optString("msg", j.optString("message", "")))
        return when {
            msg.contains("Invalid login", true) -> "Wrong email or password."
            msg.contains("already registered", true) -> "That email already has an account — sign in instead."
            msg.isNotEmpty() -> msg
            code == 429 -> "Too many attempts — wait a minute and try again."
            else -> "Something went wrong (HTTP $code)."
        }
    }

    private fun applySession(j: JSONObject) {
        accessToken = j.optString("access_token", accessToken ?: "")
        val rt = j.optString("refresh_token", "")
        if (rt.isNotEmpty()) refreshToken = rt
        expiresAt = j.optLong("expires_at",
            System.currentTimeMillis() / 1000 + j.optLong("expires_in", 3600))
    }

    /**
     * Refresh the session if we have a refresh token. If the server explicitly
     * rejects it (revoked/expired), the session is cleared and [sessionExpired]
     * is set so the UI can bounce back to sign-in. Network errors leave the
     * current session untouched.
     */
    suspend fun refreshIfPossible(): Unit = withContext(Dispatchers.IO) {
        val rt = refreshToken ?: return@withContext
        refreshLock.withLock {
            // Another caller may have refreshed while we waited on the lock.
            if (rt != refreshToken) return@withLock
            val body = JSONObject().put("refresh_token", rt).toString()
            val req = Request.Builder().url("$baseURL/auth/v1/token?grant_type=refresh_token")
                .addHeader("apikey", anonKey).addHeader("Content-Type", "application/json")
                .post(body.toRequestBody(JSON)).build()
            try {
                http.newCall(req).execute().use { r ->
                    val txt = r.body?.string() ?: "{}"
                    val j = runCatching { JSONObject(txt) }.getOrDefault(JSONObject())
                    if (r.isSuccessful && j.optString("access_token", "").isNotEmpty()) {
                        applySession(j); persist()
                    } else if (r.code in 400..403) {
                        // Refresh token rejected — this session can't recover.
                        accessToken = null; refreshToken = null; expiresAt = 0L
                        sessionExpired = true; persist()
                    }
                }
            } catch (_: IOException) { /* offline — keep the session and retry later */ }
        }
    }

    /** Refresh proactively when the token is expired or about to expire. */
    private suspend fun ensureFresh() {
        if (accessToken == null || refreshToken == null) return
        if (expiresAt == 0L || System.currentTimeMillis() / 1000 < expiresAt - 120) return
        refreshIfPossible()
    }

    /**
     * Execute a REST/Storage request with a fresh token; on 401 refresh and
     * retry once; on network failure retry once. Returns (statusCode, body)
     * or null when the network is unreachable.
     */
    private suspend fun execute(build: () -> Request): Pair<Int, String>? = withContext(Dispatchers.IO) {
        ensureFresh()
        var networkRetried = false
        var authRetried = false
        while (true) {
            try {
                http.newCall(build()).execute().use { r ->
                    val txt = r.body?.string() ?: ""
                    if (r.code == 401 && !authRetried && refreshToken != null) {
                        authRetried = true
                        refreshIfPossible()
                        if (sessionExpired) return@withContext 401 to txt
                    } else {
                        return@withContext r.code to txt
                    }
                }
            } catch (e: IOException) {
                if (networkRetried) return@withContext null
                networkRetried = true
            }
        }
        @Suppress("UNREACHABLE_CODE") null
    }

    // MARK: PostgREST

    /** Back-compat select: empty array on any failure. */
    suspend fun select(path: String): JSONArray = selectOrNull(path) ?: JSONArray()

    /** Select that distinguishes failure (null) from genuinely empty results. */
    suspend fun selectOrNull(path: String): JSONArray? {
        val res = execute {
            Request.Builder().url("$baseURL/rest/v1/$path")
                .addHeader("apikey", anonKey).addHeader("Authorization", authHeader()).get().build()
        } ?: return null
        if (res.first !in 200..299) return null
        return runCatching { JSONArray(res.second) }.getOrNull()
    }

    /**
     * Insert a row. Returns the representation array on success (empty when
     * returning=false), or null when the write did NOT reach the database.
     */
    suspend fun insert(table: String, row: JSONObject, returning: Boolean = true): JSONArray? {
        val res = execute {
            Request.Builder().url("$baseURL/rest/v1/$table")
                .addHeader("apikey", anonKey).addHeader("Authorization", authHeader())
                .addHeader("Content-Type", "application/json")
                .addHeader("Prefer", if (returning) "return=representation" else "return=minimal")
                .post(row.toString().toRequestBody(JSON)).build()
        } ?: return null
        if (res.first !in 200..299) return null
        return runCatching { JSONArray(res.second) }.getOrDefault(JSONArray())
    }

    /** Patch rows; true only if the database accepted the update. */
    suspend fun update(path: String, patch: JSONObject): Boolean {
        val res = execute {
            Request.Builder().url("$baseURL/rest/v1/$path")
                .addHeader("apikey", anonKey).addHeader("Authorization", authHeader())
                .addHeader("Content-Type", "application/json").addHeader("Prefer", "return=minimal")
                .patch(patch.toString().toRequestBody(JSON)).build()
        } ?: return false
        return res.first in 200..299
    }

    /** Delete rows; true only if the database accepted the delete. */
    suspend fun delete(path: String): Boolean {
        val res = execute {
            Request.Builder().url("$baseURL/rest/v1/$path")
                .addHeader("apikey", anonKey).addHeader("Authorization", authHeader())
                .delete().build()
        } ?: return false
        return res.first in 200..299
    }

    /** Upload bytes to Storage 'uploads' bucket, return public URL. */
    suspend fun uploadFile(bytes: ByteArray, ext: String, contentType: String): String? {
        val name = "and_${System.currentTimeMillis()}_${(0..99999).random()}.$ext"
        val res = execute {
            Request.Builder().url("$baseURL/storage/v1/object/uploads/$name")
                .addHeader("apikey", anonKey).addHeader("Authorization", authHeader())
                .addHeader("Content-Type", contentType)
                .post(bytes.toRequestBody(contentType.toMediaType())).build()
        } ?: return null
        if (res.first !in 200..299) return null
        return "$baseURL/storage/v1/object/public/uploads/$name"
    }
}
